internal/observability/metrics.go

```diff
@@ -0,0 +1,57 @@
+package observability
+
+import (
+	"context"
+	"sync"
+)
+
+type contextKey string
+
+const keyMetricsRecorder contextKey = "metricsRecorder"
+
+type (
+	Metric interface {
+		IsMetric()
+	}
+
+	MetricsRecorder interface {
+		Metrics() []Metric
+		RecordMetric(m Metric)
+	}
+)
+
+type recorder struct {
```

Comment:
I would move this into `worker`

Reply:
ended up moving everything into the `slab` package